common/persistence/cassandra/cassandraPersistence.go

```diff
@@ -798,153 +794,538 @@ func (d *cassandraPersistence) CreateWorkflowExecution(
+			// No explicit property needs to be set
+		default:
+			return nil, &types.InternalServiceError{
+				Message: fmt.Sprintf("Unknown transfer type: %v", task.GetType()),
+			}
+		}
+		t := &nosqlplugin.TransferTask{
+			Type:                    task.GetType(),
+			DomainID:                domainID,
+			WorkflowID:              workflowID,
+			RunID:                   runID,
+			VisibilityTimestamp:     task.GetVisibilityTimestamp(),
+			TaskID:                  task.GetTaskID(),
+			TargetDomainID:          targetDomainID,
+			TargetWorkflowID:        targetWorkflowID,
+			TargetRunID:             targetRunID,
+			TargetChildWorkflowOnly: targetChildWorkflowOnly,
+			TaskList:                taskList,
+			ScheduleID:              scheduleID,
+			RecordVisibility:        recordVisibility,
+			Version:                 task.GetVersion(),
+		}
+		tasks = append(tasks, t)
+	}
+	return tasks, nil
+}
+
+func (d *cassandraPersistence) prepareActivityInfosForWorkflowTxn(
+	activityInfos []*p.InternalActivityInfo,
+) (map[int64]*p.InternalActivityInfo, error) {
+	m := map[int64]*p.InternalActivityInfo{}
+	for _, a := range activityInfos {
+		_, scheduleEncoding := p.FromDataBlob(a.ScheduledEvent)
+		_, startEncoding := p.FromDataBlob(a.StartedEvent)
+		if a.StartedEvent != nil && scheduleEncoding != startEncoding {
+			return nil, p.NewCadenceSerializationError(fmt.Sprintf("expect to have the same encoding, but %v != %v", scheduleEncoding, startEncoding))
+		}
+		a.ScheduledEvent = a.ScheduledEvent.ToNilSafeDataBlob()
+		a.StartedEvent = a.ScheduledEvent.ToNilSafeDataBlob()
+		m[a.ScheduleID] = a
+	}
+	return m, nil
+}
+
+func (d *cassandraPersistence) prepareTimerInfosForWorkflowTxn(
+	timerInfo []*p.TimerInfo,
+) (map[string]*p.TimerInfo, error) {
+	m := map[string]*p.TimerInfo{}
+	for _, a := range timerInfo {
+		m[a.TimerID] = a
+	}
+	return m, nil
+}
+
+func (d *cassandraPersistence) prepareChildWFInfosForWorkflowTxn(
+	childWFInfos []*p.InternalChildExecutionInfo,
+) (map[int64]*p.InternalChildExecutionInfo, error) {
+	m := map[int64]*p.InternalChildExecutionInfo{}
+	for _, c := range childWFInfos {
+		_, initiatedEncoding := p.FromDataBlob(c.InitiatedEvent)
+		_, startEncoding := p.FromDataBlob(c.StartedEvent)
+		if c.StartedEvent != nil && initiatedEncoding != startEncoding {
+			return nil, p.NewCadenceSerializationError(fmt.Sprintf("expect to have the same encoding, but %v != %v", initiatedEncoding, startEncoding))
+		}
+
+		if c.StartedRunID == "" {
+			c.StartedRunID = emptyRunID
+		}
+
+		c.InitiatedEvent = c.InitiatedEvent.ToNilSafeDataBlob()
+		c.StartedEvent = c.StartedEvent.ToNilSafeDataBlob()
+		m[c.InitiatedID] = c
+	}
+	return m, nil
+}
+
+func (d *cassandraPersistence) prepareRequestCancelsForWorkflowTxn(
+	requestCancels []*p.RequestCancelInfo,
+) (map[int64]*p.RequestCancelInfo, error) {
+	m := map[int64]*p.RequestCancelInfo{}
+	for _, c := range requestCancels {
+		m[c.InitiatedID] = c
+	}
```

Comment:
Why reason for converting it to a map here? Looks like the keys are also not used later. For dedup?

Reply:
That's mainly for other nosqlplugin:
1. It's by nature a map, so passing a map makes more sense. And like you said, the plugin won't need to do any tricky dedup. (they might not know whether or not they should do it)
2. I am planning to let some plugin like DynamoDB to store the entire map as a blob, instead of using separate table(which is equivalent to Cassandra's non-frozen map). So passing map will make this a lot easier.